### DISCUSSION TITLE
11.0.0 Preview 7

### DISCUSSION THREAD

**Discussion by danwalmsley** [AvaloniaUI/Avalonia#11117] [General]
<hr /><em>This discussion was created from the release <a href='https://github.com/AvaloniaUI/Avalonia/releases/tag/11.0.0-preview7'>11.0.0 Preview 7</a>.</em>

Hey all, we would like to thank you again for your patience awaiting the 11.0 release! It will be worth it! We are getting close now!

I have just released preview 7

https://www.nuget.org/packages/Avalonia/11.0.0-preview7

Would like to encourage as many as possible to test this and maybe just maybe 🤞the next one might be an RC :D (no promises)

Updated list of breaking changes: https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes
Or list of breaking change PRs from this preview release: https://github.com/AvaloniaUI/Avalonia/pulls?q=is%3Apr+is%3Aclosed+label%3Abreaking-change+closed%3A%3E2023-03-21

**Comment by IsaacMarovitz** [comment]
How do we access `_topLevel.PlatformImpl.RenderScaling` in Preview 7? It seems to have become private.

**Reply by IsaacMarovitz** [reply]
Fixed with `((IRenderRoot)_topLevel).RenderScaling`

**Reply by timunie** [reply]
Note: will be added back in #11124

**Comment by rabbitism** [comment]
Any replacement of `DrawingContext.DrawBitmap`?

**Reply by maxkatz6** [reply]
DrawImage

**Reply by maxkatz6** [reply]
I.e., Bitmap implements IImage interface.

**Comment by maxkatz6** [comment]
Updated list of breaking changes, and included them in the discussion description.

**Comment by pixsperdavid** [comment]
`IHeadlessWindow` is now internal. I'm using this in order to call `GetLastRenderedFrame()` and send the output to a USB device. Is there an alternative?

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/67d68ae06df1f44647f038412d95191118654d59/src/Headless/Avalonia.Headless/HeadlessWindowExtensions.cs#L20

**Reply by pixsperdavid** [reply]
Thanks, that works, but I cannot call `IReadableBitmapImpl.Lock()` on the bitmap as the implementation is now also internal.

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/31a691c55032e5be69776f4279612a516c415bcc/src/Avalonia.Base/Media/Imaging/Bitmap.cs#L207

**Reply by pixsperdavid** [reply]
I understand this will work, but this is not the same thing. I want to obtain a reference to the framebuffer in order to pass the memory location to another function (in this case an umanaged pinvoke call). Adding a copy will negatively affect performance.

**Reply by maxkatz6** [reply]
Lock won't be returned, as there is zero guarantee that bitmap is in actual RAM (it could be in GPU memory).
Technically, CaptureRenderedFrame can return lockable WriteableBitmap in the next preview, an issue should be created for that.

**Reply by pixsperdavid** [reply]
Ok thanks, issue created #11130

**Comment by KieranDevvs** [comment]
Ive just upgraded from preview 4 and im getting an error when trying to publish a trimmed (link mode) native AOT binary.
```txt
MainWindow.axaml.cs(9,13,9,32): error CS1501: No overload for method 'InitializeComponent' takes 2 arguments
```
I thought it was to do with trimming but i've disabled it and i'm still getting the error.

Should I open an issue for this?

**Reply by maxkatz6** [reply]
No, please read breaking changes. Since you upgraded from preview4 to preview7 see the full list.
You need to remove XamlNameReferenceGenerator package from your projects.

**Reply by KieranDevvs** [reply]
Yeah i've done all of that, I can run the build as debug, I just can't publish.
It appears that it has broken in preview 6 as 5 works as expected.

Here are my package references:
```xml

	<ItemGroup>
		<PackageReference Include="Avalonia" Version="11.0.0-preview7" />
		<PackageReference Include="Avalonia.Desktop" Version="11.0.0-preview7" />
		<PackageReference Include="Avalonia.Themes.Fluent" Version="11.0.0-preview7" />
		<!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
		<PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.0.0-preview7" />
		<PackageReference Include="Avalonia.ReactiveUI" Version="11.0.0-preview7" />
	</ItemGroup>
```

**Reply by KieranDevvs** [reply]
Its publishing successfully now after I downgraded and upgraded.
Maybe VS was caching something? Thanks anyway.

**Comment by kyuranger** [comment]
![image](https://user-images.githubusercontent.com/30720036/234440478-e1344e5a-198b-4bf9-96cb-57ca7c0fe641.png)

Visual Studio Template is still staying at Preview 5.0 and I have to upgrade the NuGet package every time while creating a new project.

**Comment by codecat** [comment]
After updating from Preview4 (and going through the breaking changes list) I am now getting this (rather vague) compilation error:

```
App.axaml : Avalonia error AVLN:0004: Object reference not set to an instance of an object.
```

How can I go about troubleshooting this?

**Edit:** This seems to be related to #10046.

**Comment by grabthar** [comment]
High CPU usage for the axaml previewer in Visual studio 2022
(Persistant CPU usage of the associated .NET host)

Repro steps:
Create a new Avalonia .NET Core App
Upgrade to preview7 (remove XamlNameReferenceGenerator)
Build
Open MainWindow.axaml then look your cpu usage.

**Reply by codecat** [reply]
I can confirm this, though this should probably be an issue, not part of discussions?

**Reply by grabthar** [reply]
Sorry for the wrong place + this is a known issue
https://github.com/AvaloniaUI/Avalonia/issues/10203
